sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,133 @@
+import { EventEmitter } from "events";
+import { delay, HttpOperationResponse, RequestOptionsBase, ServiceClient } from "@azure/core-http";
+
+export interface PollerOptionalParameters<States> {
+  manual?: boolean;
+  intervalInMs?: number;
+  state?: States;
+  client?: ServiceClient;
+  initialResponse?: HttpOperationResponse;
+  previousResponse?: HttpOperationResponse;
+  requestOptions?: RequestOptionsBase;
+  stopped?: boolean;
+}
+
+export abstract class Poller<States, Result> extends EventEmitter  {
+  private readonly manual: boolean = false;
+  private stopped: boolean = false;
+  private intervalInMs: number = 1000;
+  private _state: States;
+  protected readonly client: ServiceClient;
+  protected initialResponse?: HttpOperationResponse;
+  protected previousResponse?: HttpOperationResponse;
+  protected requestOptions: RequestOptionsBase | undefined;
+  public stopped: boolean;
+
+  constructor(options: PollerOptionalParameters) {
+    this.manual = Boolean(options.manual);
+    this.intervalInMs = Number(options.intervalInMs);
+    this.client = options.client;
+    this.initialResponse = options.initialResponse;
+    this.previousResponse = options.previousResponse;
+    this.requestOptions = options.requestOptions;
+    this.stopped = Boolean(options.stopped);
+
+    if (options.state) this._state = options.state;
+    if (this.manual) return;
+    this.poll();
+  }
+
+  protected abstract getStateFromResponse(
+    response: HttpOperationResponse
+  ): LongRunningOperationStates;
+
+  protected abstract async sendRequest(options?: RequestOptionsBase): Promise<HttpOperationResponse>;
+
+  protected async processResponse(response: HttpOperationResponse): Promise<void | Result> {
```

Comment:
this method feels weird, the promise only has a non-empty result when it's the final request and the only time that result gets used is when manually polling with pollOnce().

Reply:
What pattern would you prefer for this?

Reply:
I think having the promise always void and if someone wants the final value they can grab it later. Otherwise it'll be very confusing to distinguish between the result being undefined because that *is* the final result and undefined because it's not.

Reply:
gotcha! I'll update it